### ISSUE TITLE
Vehicle API additions, that might be good

### ISSUE THREAD

**Issue by ByAlexius** [yamp-project/v-issues#2]
**What is the scope of the api ?**
- [ ] World
- [ ] Ped
- [X] Vehicle

**Propose an api design**

I have think these additions would be great and help developers while programming with the Vehicle API.
The code examples use Java variable types for parameters.

**Note:** While I primarily described setter methods, getters are also required. However, including them might make the list to long.

| Code Example  | Functionality |
| ------------- | ------------- |
| getSeatedPed(int seatPosition/seatId) | Returns the ped seated at the specified seat ID (described in code). If no one is seated, returns null. |
| getAllSeatedPeds()  | Returns an array or list containing all peds currently inside the vehicle, along with their corresponding seat IDs (described above).  |
| eject(int seatId) | Removes the ped from the vehicle based on the provided seat ID. |
| eject(int playerId) | Ejects the player from the vehicle if they are currently inside. |
| eject(int pedId) | Ejects the specified ped from the vehicle if they are currently inside. |
| eject(Player player)  | Ejects the player from the vehicle using a direct player object reference. |
| eject(Ped ped)  | Ejects the specified ped from the vehicle using a direct ped object reference. |
| setLockState(boolean lockState) | Sets the vehicle's lock state. If true, no ped can enter or exit (implementing the "cannot exit" part is optional). |
| destroyEngine()  | Sets the engine health to zero (completely damaged). |
| repairEngine()  | Sets the engine health to 100 (or the maximum value). |
| disableEngine()  | Disables the engine, preventing it from starting until enableEngine() is called. |
| enableEngine()  | Enables the engine, allowing it to start again (counterpart to disableEngine() |
| setLightState(int lightId, int brightness (optional))  | Allows individual light control for the car. Enables/disables each light and optionally sets its brightness (not all lights may have adjustable brightness). |
| **Please Read:** | **The following functions could be combined into a single function that accepts a JSON string containing all the car's cosmetic features. However, I believe individual methods provide better readability.** |
| setPrimaryColor(String hex)  | Sets the primary car color using a hex value. |
| setPrimaryColor(int r, int g, int b)  | Sets the primary car color using RGB values. |
| setPrimaryColor(String color)  | Sets the primary car color using predefined color names (white, black, yellow, etc.). |
| **Please Read:** | **Similar functionality can be implemented for secondary colors and the pearlescent effect.** | 
| setLicensePlate(String licensePlateText)  | Sets the license plate text. If exceeding the maximum length, truncates the text. |
| setVisibility(boolean state)  | Sets the vehicle's visibility. Consider adding an option to keep internal peds visible while hiding the vehicle itself. |

I will update this list if i think of more methods.

If you have improvement idea's, don't hesitate to add comments.

Thanks for reading, have a nice day :)

**Comment by ByAlexius** [comment]
> we appreciate your proposal, can you please somewhat refactor `eject` functions part? it would be impossible to differentiate between seat, player or ped ids

I have removed the function's that are impossible to implement and have added a ejectAll() method.
Can you take a look again?

With kind regards
Alexander

**Comment by drakeee** [comment]
> > we appreciate your proposal, can you please somewhat refactor `eject` functions part? it would be impossible to differentiate between seat, player or ped ids
> 
> I have removed the function's that are impossible to implement and have added a ejectAll() method. Can you take a look again?
> 
> With kind regards Alexander

it is possible to differentiate different types of arguments, i.e: `eject(playerObject)` would be doable along with `eject(int seatId)`

**Comment by ByAlexius** [comment]
> > > we appreciate your proposal, can you please somewhat refactor `eject` functions part? it would be impossible to differentiate between seat, player or ped ids
> > 
> > 
> > I have removed the function's that are impossible to implement and have added a ejectAll() method. Can you take a look again?
> > With kind regards Alexander
> 
> it is possible to differentiate different types of arguments, i.e: `eject(playerObject)` would be doable along with `eject(int seatId)`

Ok, I added back `eject(Player player)`, if i understood correctly this should work.
Are there any other problems or just this?

**Comment by drakeee** [comment]
> > > > we appreciate your proposal, can you please somewhat refactor `eject` functions part? it would be impossible to differentiate between seat, player or ped ids
> > > 
> > > 
> > > I have removed the function's that are impossible to implement and have added a ejectAll() method. Can you take a look again?
> > > With kind regards Alexander
> > 
> > 
> > it is possible to differentiate different types of arguments, i.e: `eject(playerObject)` would be doable along with `eject(int seatId)`
> 
> Ok, I added back `eject(Player player)`, if i understood correctly this should work. Are there any other problems or just this?

yeah, sorry if it was gibberish, but that was exactly what I meant, other functions looks good

**Comment by ByAlexius** [comment]
Quick Addition:

`getPosition();`
and
`getDimension();`

**Comment by devpanda0** [comment]
> Quick Addition:
> 
> `getPosition();` and `getDimension();`

i think that has every entity later

**Comment by ByAlexius** [comment]
> > Quick Addition:
> > `getPosition();` and `getDimension();`
> 
> i think that has every entity later

Yeah, probably but writing something that is already planned is better then not writing it down.
If you understand what i mean.

**Comment by szkiddaj** [comment]
In my opinion, a lot of these should be a getter/setter instead of methods, because they don't require more than 1 argument. So instead of `.getDimension()`, a `.dim` or `.dimension` getter would be better.

**Comment by LeonMrBonnie** [comment]
> In my opinion, a lot of these should be a getter/setter instead of methods, because they don't require more than 1 argument. So instead of `.getDimension()`, a `.dim` or `.dimension` getter would be better.

Every method that has a get and a set that only accepts / returns one value, should be a property.

**Comment by ByAlexius** [comment]
Since you are probably talking about javascript/typescript i can't say much since i have very limited experience using it.

**Comment by LeonMrBonnie** [comment]
The same can be applied to any language that supports getters and setters

**Comment by ByAlexius** [comment]
I based my original code example of java, where i generally use get and set methods (generated by lombok) for variables if it is used outside it's own class, i generally use the variable directly only if the method is in the same class

**Comment by LeonMrBonnie** [comment]
Nothing about the API here should be discussed in regards to Java because that is a dying language that will never be available in YAMP

**Comment by ByAlexius** [comment]
> Nothing about the API here should be discussed in regards to Java because that is a dying language that will never be available in YAMP

Yeah fair enough (I used it for the example code since my main program language for work is java and I also didn't want to write wrong c++ or javascript code).

**Comment by kyro95** [comment]
```
destroyEngine( | Sets the engine health to zero (completely damaged).
repairEngine() | Sets the engine health to 100 (or the maximum value).
disableEngine() | Disables the engine, preventing it from starting until enableEngine() is called.
enableEngine() | Enables the engine, allowing it to start again (counterpart to disableEngine()
```

to -> 
```
setEngineHealth(health: number);
setEngineState(state: bool)
```

**Comment by ByAlexius** [comment]
**Changes:**
removed:
`destroyEngine()`
`repairEngine()`
`disableEngine()`
`enableEngine()`

added:
`setEngineHealth(int health)`
`setEngineState(boolean state)`

Thanks @kyro95 for the suggestion :)
